test/Table_Tests/src/Common_Table_Operations/Join/Union_Spec.enso

```diff
@@ -245,29 +290,80 @@ run_union_tests group_builder setup call_union =
+
+        t3 = call_union [t1, t2]
+        expect_column_names ["A"] t3
+        w = Problems.expect_only_warning Loss_Of_Integer_Precision t3
+        # TODO should we try to include column name here for context? may be worth it...
+        w.affected_rows_count.should_equal 1
+        t3.at "A" . value_type . is_floating_point . should_be_true
+        t3.at "A" . to_vector . should_equal [(2^70).to_float, 2^10, 2, 1.5, 2.0, 2.5]
+
+    group_builder.specify "should find a common type (Integer and Char of different sizes)" <|
+        t1 = (table_builder [["X", [0, 1, 2]], ["Y", ['aa', 'bb', 'cc']]]) . cast "X" (Value_Type.Integer Bits.Bits_16) . cast "Y" (Value_Type.Char size=2 variable_length=False)
+        t2 = (table_builder [["X", [3, 4, 5]], ["Y", ['x', 'y', 'z']]]) . cast "X" (Value_Type.Integer Bits.Bits_32) . cast "Y" (Value_Type.Char size=1 variable_length=False)
+        supports_complex_types = (t1.is_error || t2.is_error || Problems.get_attached_warnings t1 . not_empty).not
+        case supports_complex_types of
+            False -> Nothing
+            True ->
+                t12 = call_union [t1, t2]
+                # No warnings are expected
+                Problems.assume_no_problems t12
+                t12.at "X" . value_type . should_equal (Value_Type.Integer Bits.Bits_32)
+                t12.at "Y" . value_type . should_equal (Value_Type.Char size=2 variable_length=True)
+
+                t12.at "X" . to_vector . should_equal [0, 1, 2, 3, 4, 5]
+                t12.at "Y" . to_vector . should_equal ['aa', 'bb', 'cc', 'x', 'y', 'z']
 
-            t5 = call_union [t2, t4]
-            expect_column_names ["A"] t5
-            t5.at "A" . value_type . is_floating_point . should_be_true
-            t5.at "A" . to_vector . should_equal [1.0, 2.0, 2.5, 2^100, 2^10, 2]
+    date_time_pending = if setup.test_selection.date_time.not then "Date/Time operations are not supported."
+    group_builder.specify "should warn when converting a Date to Date_Time" pending=date_time_pending <|
+        t1 = table_builder [["D", [Date_Time.new 2024 5 16 16 48 23]]]
+        t2 = table_builder [["D", [Date.new 2019 10 23, Date.new 2020]]]
 
-            t6 = call_union [t1, t4]
-            expect_column_names ["A"] t6
-            t6.at "A" . value_type . should_be_a (Value_Type.Decimal ...)
-            t6.at "A" . to_vector . should_equal [0, 1, 2, 2^100, 2^10, 2]
+        action = call_union [t1, t2] on_problems=_
+        tester table =
+            IO.println ""
+            table.print
+            expect_column_names ["D"] table
+            table.at "D" . value_type . should_equal Value_Type.Date_Time
+            table.at "D" . to_vector . should_equal_tz_agnostic [Date_Time.new 2024 5 16 16 48 23, Date_Time.new 2019 10 23 0 0 0, Date_Time.new 2020 1 1 0 0 0]
+        problems = [Mixing_Date_Time_Types.Date_To_Date_Time "D"]
+        Problems.test_problem_handling action problems tester
```

Comment:
This example shows why the warning is needed.

When running in-memory, the coercion happens as +- expected:
```
   | D
---+------------------------------------------
 0 | 2024-05-16 16:48:23+02:00[Europe/Warsaw]
 1 | 2019-10-23 00:00:00+02:00[Europe/Warsaw]
 2 | 2020-01-01 00:00:00+01:00[Europe/Warsaw]
```

But when running on Postgres, it's a bit different story:
```
 D
----------------------
 2024-05-16 14:48:23Z
 2019-10-22 22:00:00Z
 2019-12-31 23:00:00Z
```

This behaviour actually _does make sense_ - the Date is converted to Date_Time by setting it to midnight _at the timezone it was inputted in_ - which is `Europe/Warsaw` in my case. However then, when we round-trip it back from Postgres to JDBC, due to how Postgres stores the data, it is sent to us in UTC. Hence it is no longer `00:00`. To make matters worse - it is sometimes `22:00` and sometimes `23:00` - because `Europe/Warsaw` has DST but UTC doesn't! This is why a warning is useful to let the user know that things may not necessarily work as they expect.

Reply:
I guess a slightly safer way to do that could be to _always_ interpret the Date as `00:00` in UTC instead of default zone. That would make Postgres and in-memory consistent. But it may still not be a catch-all solution once we add more DB backends. This kind of coercion is bound to have some implicit assumptions, so I think best course of action is relying on backend's default and warning (current behaviour).